File: php/src/Tracker.php
<?php

declare(strict_types=1);

namespace Royaltics\ErrorTracker;

use Royaltics\ErrorTracker\Types\ClientConfig;
use Royaltics\ErrorTracker\Types\EventLevel;

final class Tracker
{
    private static array $instances = [];
    private static ?ErrorTrackerClient $defaultInstance = null;

    public static function create(ClientConfig $config, ?string $name = null): ErrorTrackerClient
    {
        $client = new ErrorTrackerClient($config);

        if ($name !== null) {
            self::$instances[$name] = $client;
        } elseif (self::$defaultInstance === null) {
            self::$defaultInstance = $client;
        }

        return $client->start();
    }

    public static function get(?string $name = null): ErrorTrackerClient
    {
        if ($name !== null) {
            if (!isset(self::$instances[$name])) {
                throw new \RuntimeException("Tracker instance \"{$name}\" not found");
            }
            return self::$instances[$name];
        }

        if (self::$defaultInstance === null) {
            throw new \RuntimeException('No default tracker initialized. Call Tracker::create() first.');
        }

        return self::$defaultInstance;
    }

    public static function error(
        \Throwable $error,
        EventLevel $level = EventLevel::ERROR,
        ?array $metadata = null
    ): ErrorTrackerClient {
        return self::get()->error($error, $level, $metadata);
    }

    public static function event(
        string $title,
        EventLevel $level = EventLevel::INFO,
        ?array $metadata = null
    ): ErrorTrackerClient {
        return self::get()->event($title, $level, $metadata);
    }

    public static function flush(): void
    {
        self::get()->forceFlush();
    }

    public static function pause(): ErrorTrackerClient
    {
        return self::get()->pause();
    }

    public static function resume(): ErrorTrackerClient
    {
        return self::get()->resume();
    }

    public static function shutdown(): void
    {
        self::$defaultInstance?->shutdown();
        
        foreach (self::$instances as $instance) {
            $instance->shutdown();
        }

        self::$instances = [];
        self::$defaultInstance = null;
    }

    public static function has(?string $name = null): bool
    {
        if ($name !== null) {
            return isset(self::$instances[$name]);
        }

        return self::$defaultInstance !== null;
    }
}

File: go/tracker.go
package errortracker

import (
	"fmt"
	"sync"

	"github.com/royaltics/tracker-go/types"
)

var (
	instances       = make(map[string]*ErrorTrackerClient)
	defaultInstance *ErrorTrackerClient
	mu              sync.RWMutex
)

func Create(config *types.ClientConfig, name ...string) (*ErrorTrackerClient, error) {
	client, err := NewClient(config)
	if err != nil {
		return nil, err
	}

	mu.Lock()
	defer mu.Unlock()

	if len(name) > 0 && name[0] != "" {
		instances[name[0]] = client
	} else if defaultInstance == nil {
		defaultInstance = client
	}

	return client.Start(), nil
}

func Get(name ...string) (*ErrorTrackerClient, error) {
	mu.RLock()
	defer mu.RUnlock()

	if len(name) > 0 && name[0] != "" {
		client, ok := instances[name[0]]
		if !ok {
			return nil, fmt.Errorf("tracker instance %q not found", name[0])
		}
		return client, nil
	}

	if defaultInstance == nil {
		return nil, fmt.Errorf("no default tracker initialized. Call Create() first")
	}

	return defaultInstance, nil
}

func Error(err error, level types.EventLevel, metadata map[string]string) error {
	client, e := Get()
	if e != nil {
		return e
	}
	client.Error(err, level, metadata)
	return nil
}

func Event(title string, level types.EventLevel, metadata map[string]string) error {
	client, err := Get()
	if err != nil {
		return err
	}
	client.Event(title, level, metadata)
	return nil
}

func Flush() error {
	client, err := Get()
	if err != nil {
		return err
	}
	return client.ForceFlush()
}

func Pause() error {
	client, err := Get()
	if err != nil {
		return err
	}
	client.Pause()
	return nil
}

func Resume() error {
	client, err := Get()
	if err != nil {
		return err
	}
	client.Resume()
	return nil
}

func Shutdown() error {
	mu.Lock()
	defer mu.Unlock()

	if defaultInstance != nil {
		if err := defaultInstance.Shutdown(); err != nil {
			return err
		}
		defaultInstance = nil
	}

	for _, client := range instances {
		if err := client.Shutdown(); err != nil {
			return err
		}
	}

	instances = make(map[string]*ErrorTrackerClient)
	return nil
}

func Has(name ...string) bool {
	mu.RLock()
	defer mu.RUnlock()

	if len(name) > 0 && name[0] != "" {
		_, ok := instances[name[0]]
		return ok
	}

	return defaultInstance != nil
}
